src/dispatch/builder.rs

```diff
@@ -0,0 +1,209 @@
+use smallvec::SmallVec;
+
+use dispatch::{Dispatcher, SystemDependencies, SystemId, SystemInfo, ThreadLocal};
+use dispatch::stage::StagesBuilder;
+use system::{System, SystemData};
+
+#[derive(Default)]
+pub struct IdGen {
+    current: usize,
+}
+
+impl IdGen {
+    pub fn next(&mut self) -> SystemId {
+        let id = self.current;
+        self.current += 1;
+
+        SystemId(id)
+    }
+}
+
+/// Builder for the [`Dispatcher`].
+///
+/// [`Dispatcher`]: struct.Dispatcher.html
+///
+/// ## Barriers
+///
+/// Barriers are a way of sequentializing parts of
+/// the system execution. See `add_barrier()`.
+///
+/// ## Examples
+///
+/// This is how you create a dispatcher with
+/// a shared thread pool:
+///
+/// ```rust
+/// # #![allow(unused)]
+/// #
+/// # extern crate shred;
+/// # #[macro_use]
+/// # extern crate shred_derive;
+/// # use shred::{Dispatcher, DispatcherBuilder, Fetch, System};
+/// # #[derive(Debug)] struct Res;
+/// # #[derive(SystemData)] #[allow(unused)] struct Data<'a> { a: Fetch<'a, Res> }
+/// # struct Dummy;
+/// # impl<'a> System<'a> for Dummy {
+/// #   type SystemData = Data<'a>;
+/// #
+/// #   fn run(&mut self, _: Data<'a>) {}
+/// # }
+/// #
+/// # fn main() {
+/// # let system_a = Dummy;
+/// # let system_b = Dummy;
+/// # let system_c = Dummy;
+/// # let system_d = Dummy;
+/// # let system_e = Dummy;
+/// let dispatcher: Dispatcher = DispatcherBuilder::new()
+///     .add(system_a, "a", &[])
+///     .add(system_b, "b", &["a"]) // b depends on a
+///     .add(system_c, "c", &["a"]) // c also depends on a
+///     .add(system_d, "d", &[])
+///     .add(system_e, "e", &["c", "d"]) // e executes after c and d are finished
+///     .build();
+/// # }
+/// ```
+///
+#[derive(Default)]
+pub struct DispatcherBuilder<'a, 'b> {
+    id_gen: IdGen,
+    map: FnvHashMap<String, SystemId>,
+    stages_builder: StagesBuilder<'a>,
+    thread_local: ThreadLocal<'b>,
+    #[cfg(not(target_os = "emscripten"))]
+    thread_pool: Option<::std::sync::Arc<::rayon_core::ThreadPool>>,
+}
+
+impl<'a, 'b> DispatcherBuilder<'a, 'b> {
+    /// Creates a new `DispatcherBuilder` by using the `Default` implementation.
+    ///
+    /// The default behaviour is to create a thread pool on `finish`.
+    /// If you already have a rayon `ThreadPool`, it's highly recommended to configure
+    /// this builder to use it with `with_pool` instead.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Adds a new system with a given name and a list of dependencies.
+    /// Please not that the dependency should be added before
+    /// you add the depending system.
+    ///
+    /// # Panics
+    ///
+    /// * if the specified dependency does not exist
+    pub fn add<T>(mut self, system: T, name: &str, dep: &[&str]) -> Self
+        where T: for<'c> System<'c> + Send + 'a
+    {
+        let id = self.id_gen.next();
+        let mut reads: Vec<_> = T::SystemData::reads(0);
+        let writes = T::SystemData::writes(0);
```

Comment:
Thanks for reviewing, @WaDelma!

My code is correct, however it's not obvious why I did it like that, because I never wrote my design decision down.

First of all, in case you don't know, system data most likely consists of other structs implementing `SystemData`, so for example: 

```rust
#[derive(SystemData)]
struct Data<'a> {
    a: ReadStorage<'a, A>,
    b: WriteStorage<'a, B>,
}
```

Now, I'd like resources to have additional ids (in Specs these were called component ids). So `fetch` / `reads` / `writes` should be parameterized over this resource id. However, the system data directly associated with a system, doesn't need such an id. It just gets passed `0` for that reason.

> Shouldn't these 0s be id instead?

The `id` you're seeing here is the system id, which is totally unrelated to the resource id.